File: 백준/Gold/1744. 수 묶기/수 묶기.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

public class Main {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int N = Integer.parseInt(br.readLine());

        ArrayList<Integer> minus = new ArrayList<>();
        List<Integer> plus = new ArrayList<>();


        for (int i = 0; i < N; i++) {
            int tmp = Integer.parseInt(br.readLine());
            if (tmp <= 0){
                minus.add(tmp);
            } else {
                plus.add(tmp);
            }
        }

        minus.sort(Comparator.naturalOrder());
        plus.sort(Comparator.reverseOrder());
        int result = 0;
        for (int i = 0; i < minus.size(); i+=2) {
            if (i == minus.size()-1) {
                result += minus.get(i);
            } else {
                result += (minus.get(i) * minus.get(i+1));
            }
        }

        for (int i = 0; i < plus.size(); i+=2) {
            if (i == plus.size() - 1) {
                result += plus.get(i);
            } else {
                int multi = plus.get(i) * plus.get(i+1);
                int sum = plus.get(i) + plus.get(i+1);
                int tmp = 0;
                if (multi > sum) {
                    tmp = multi;
                } else{
                    tmp = sum;
                }
                result += tmp;
            }
        }

        System.out.println(result);



        
    }
}

File: 백준/Gold/1744. 수 묶기/수 묶기.py
import sys
input = sys.stdin.readline

# N 최대 50
# 모든 경우의 수 탐색시
# 50C2 * 48C2 *

#정렬 nlogn
# 그후 n에 처리

# -10 -5 -4 -1 0 5 10 110
# 정렬 후
# 일단 음수 ~ 0 리스트
# 양수 리스트

#예외 사항 -> 중복해서 들어올 수 있다.
# 1 1 

# -5 -4 -1 -1 -1 0 1 1 1 5 5 

N = int(input())

minus = []
plus = []

for _ in range(N):
    tmp = int(input())

    if tmp <= 0:
        minus.append(tmp)
    else:
        plus.append(tmp)


minus.sort()
plus.sort(reverse=True)

result = 0


for i in range(0,len(minus),2):

    # print(f'i {i}')

    #홀수로 minus가 남는다면
    if i == len(minus) - 1:
        result += minus[i]
    else:
        result += (minus[i] * minus[i+1])

for i in range(0,len(plus),2):

    if i == len(plus) -1:
        result += plus[i]
    else:
        곱 = plus[i] * plus[i+1]
        합 = plus[i] + plus[i+1]
        결과 = 0
        if 곱 > 합:
            결과 = 곱
        else:
            결과 = 합
        result += (결과)

# print(minus)
# print(plus)

print(result)

